README.md
![Script](https://img.shields.io/badge/WaXxX-G2ETH-BF5B16?style=plastic&logo=Ethereum)
![ETH](https://img.shields.io/badge/Ethereum-3C3C3D?style=plastic&logo=Ethereum&logoColor=orange)
![Python](https://img.shields.io/badge/Python-FFD43B?style=plastic&logo=python&logoColor=darkgreen)
![g](https://img.shields.io/badge/GitHub-%2312100E.svg?&style=plastic&logo=Github&logoColor=BF5B16&?labelColor=BF5B16)
![Blah](https://img.shields.io/badge/Python3-RE-7A297B?style=plastic&logo=Python)
![fff](https://img.shields.io/badge/Python3-Requests-7A297B?style=plastic&logo=Python)



# ***G2ETH***
<p align="center">
  <!--<img width="300" height="300" src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/python.svg">-->
  <img width="300" height="300" src="https://imgur.com/KHidht1.png">
</p>

<hr>

### Get ETH mining stats on a given GPU
##### **Returns current Bitcoin price, current ***ETH*** price, ***ETH*** hashrate of the given GPU, daily, weekly and monthly estimated profit, year the GPU was released, ROI(Return of Investment), Power consumption, efficiency and TDP.**  **Lists all cards capable of mining ***ETH*** and their ***ETH*** mining stats.** ***g2eth*** **also has a USD/GBP/CAD/EUR to ETH converter and vice versa.**. **Check the current ETH and BTC prices.**
##### 

##### Available statistics:
* Get mining stats on GPUs
  *  Get ETH hash rate on a supplied GPU
  *  Get current ETH price
  *  Get current BTC price
  *  Daily, weekly and monthly estimated earnings
  *  MSRP, Release year, projected ROI, TDP, Efficiency, Power consumption
* List all GPUs capable of mining ETH
* Convert ETH to USD/GBP/CAD/EUR and vice versa (USD/CAD/GBP/EUR to ETH)
* Get current ETH and BTC price and status


<hr>

##### To install:
```python
git clone --depth 1 https://github.com/Waxxx333/g2eth.git
cd g2eth
chmod +x install.sh
┌──[яoot@ᴡɪɴᴅᴏᴡꜱ95]-<g2eth>
└─⋗ ./install.sh
```
##### Or simply run it:
```python
git clone --depth 1 https://github.com/Waxxx333/g2eth
cd g2eth
chmod +x g2eth
┌──[яoot@ᴡɪɴᴅᴏᴡꜱ95]-<g2eth>
└─⋗ python3 g2eth -h/--help
g2eth.py GPU ETH mining stats

options:
  -g GPU, --gpu GPU     GPU to search for
  -h, --help            Show this help menu
  -u, --usage           Advanced Usage
  -l, --list            List all cards capable of mining ETH
  -p, --price           Show BTC and ETH price
  -c CONVERT, --convert CONVERT | Currency to conver ETH/USD/GBP
  -i INTO, --into INTO | Currency to convert into ETH/USD/GBP
  -n AMOUNT, --amount AMOUNT | Amount to convert
```
#### Command formatting: 
 - <kbd>g2eth -g 3060ti</kbd> = 3060 Ti Non-LHR
 - <kbd>g2eth -g 3060tilhr</kbd> = 3060 Ti LHR 
 - <kbd>g2eth -g 3060lhr</kbd> = 3060 LHR 
 - <kbd>g2eth -g 3060</kbd> = 3060 Non-LHR 
 - <kbd>g2eth -g 2060s</kbd> Super cards (1660 Super | 2060 Super) 
 - Etc. . .

##### Example output:
```python
┌──[яoot@ᴡɪɴᴅᴏᴡꜱ95]-<g2eth>
└─⋗ g2eth -g 3060tilhr
[■] NVIDIA RTX 3080 LHR [■]
[■] BTC Price: $37,280.0200 .::. ETH Price: $2448 .::. Up: 2.7543
[■] ETH Hashrate(DaggerHashimoto): 71.70 Mh/s
[■] 24h: $2.15 / Weekly: $15.05 / Monthly: $64.50
[■] Released: 2021 .::. MSRP: $999
[■] ROI: 465 days .::. Efficiency 0.316 Mh/w
[■] TDP: 320 watt .::. Power: 227w
``` 
 ##### ***Using the*** <kbd>-l/--list</kbd> ***flag will show all cards capable of mining <kbd>ETH</kbd> and it will also give the name you need to use in the command to search the stats***
###### <kbd>-l/--list</kbd> arg will give something like the following. The name after <kbd>Command name:</kbd> is the name you'll use to search the GPU by.
```python
┌──[яoot@ᴡɪɴᴅᴏᴡꜱ95]-<g2eth>
└─⋗ ./g2eth.py -l
[■] Card name: NVIDIA GTX 1080 .:|:. Command name: 1080
[■] Card name: AMD RX 6800 .:|:. Command name: 6800
[■] Card name: NVIDIA CMP 90HX .:|:. Command name: cmp90
[■] Card name: NVIDIA P106-100 .:|:. Command name: p106
```
##### Convert USD/GBP/CAD/EUR to ETH or ETH to USD/GBP/CAD/EUR <img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/ethereum.svg#gh-light-mode-only" alt="python" align=left width=24><img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/ethereum.svg#gh-dark-mode-only" alt="python" align=left width=24><br>
```python
g2eth --convert/-c [USD/GBP/CAD/EUR/ETH] --into/-i [USD/GBP/CAD/EUR/ETH] --amount/-n [AMOUNT]
┌──[яoot@ᴡɪɴᴅᴏᴡꜱ95]-<g2eth>
└─⋗ g2eth.py --convert eth --into usd --amount 20
[■] Converting: ♦20.0 ETH into: $48508.72 USD [■]
```
<hr>

###### Requirements: <kbd>python-requests</kbd> | but the installation script will attempt to install it. 
<!--
|Feature            |Termux | Linux | Windows | WSL
|-------------------|-----|-------|---|------------|
| Installer Compatible with    |    |✓      |  |   ✓-->
###### **For** ***Windows*** and ***Termux*** **you will have to manually install python-requests**. **You will need <kbd>pip</kbd> if you're on Windows or Termux to install <kbd>requests</kbd>.** **I will be fixing this at some point. Right now the installer supports: **openSUSE, Arch-based distros, Debian-based distros and Fedora**
###### **This script doesn't work for ***GTX 1050*** cards as because you need ***<kbd>>4GB</kbd>*** ***VRAM*** to mine ***ETH*****. 

###### G2ETH now has tab completion if you install it via <kbd>install.sh</kbd>. You must be using <kbd>zsh</kbd> or <kbd>bash</kbd> and you also need to have <kbd>bash-completion</kbd> for <kbd>bash</kbd> or <kbd>zsh-completions</kbd> for <kbd>zsh</kbd>. Run the install script, close your shell, reopen a shell and type <kbd>g2eth --</kbd> (***two hyphens***) and press ***tab***.


<hr>

<h1 align="center">Screenshots</h1>

<p align="center">
  <img src="https://i.imgur.com/Dt8Q56c.png" width="670" height="900">
  <img src="https://i.imgur.com/SxDR9aQ.png" width="670" height="500">
</p>


<!--
* <kbd>1.3</kbd>
    * Updated code accordingly to source's code change
-->

<details>
  <summary><kbd>Changes</kbd></summary>
  <ul>
    <li><b>Added function to convert ETH to USD/GBP and vice versa</li>
    <li><b>Added function to list ALL cards and all of their stats</li>
    <li><b>Shows <kbd>+/-</kbd> ETH</li>
    <li>Added price function</li>
    <li>Added CAD and EUR to the converter</li>
    <li>Updated script according to the source's source code change(Source: let me know if there's a problem with this project)</li>
    <li>Fixed conversion decimal problem</li>
    <li>Got rid of redundant "list all and stat" function</li>
    <li>Added zsh and bash tab completion</li>
    </ul>
</details>

> To Do
- [ ] Add more distros' package managers to the installer
- [ ] Make installer work in Winblows
- [ ] Make installer work with Termux
- [ ] Make STDOUT <kbd>prettier</kbd> on Termux
- [ ] Maybe a side-by-side comparison function
  - [x] <strike>***Add current ETH price function***</strike>
  - [x] <strike>***Add Canadian Dollars***</strike>
  - [x] <strike>***Make <kbd>install.sh</kbd>***</strike> 
  - [x] <strike>***Create <kbd>requirements.txt</kbd>***</strike>
  - [x] Work on bash and zsh completion

##### GPU stats are from [hashrate.no](https://hashrate.no)

  
<hr>

<p align="center">
  <img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/ethereum.svg" width="75" height="75">
  <img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/python.svg" width="75" height="75">
  <img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/amd.svg" width="75" height="75">
  <img src="https://raw.githubusercontent.com/simple-icons/simple-icons/develop/icons/nvidia.svg" width="75" height="75">
</p>
<p align="center">
  <a href="https://twitter.com/waxxx333"><img src="https://img.shields.io/badge/-WaXxX-E34F26?style=plastic&logo=Windows%2095&logoColor=white"></a>
</p>

<hr>


CHANGELOG.md
Added Version
Added CAD to converter
Added current price function
Added EUR
Added check list for supported currencies
02-02-22--11:41
Modified for website source code change && fixed price function


g2eth.py
#!/bin/python3
import requests, re, argparse, os, sys, random;
from time import sleep;
from decimal import Decimal as D;
DB=("\033[01;38;5;21m")
RD=("\033[01;38;5;9m")
PNK=("\033[01;38;5;201m")
PRP=("\033[01;38;5;55m")
GRN=("\033[01;38;5;10m")
DRK=("\033[01;38;5;242m")
WHT=("\033[01;38;5;15m")
ORN=("\033[01;38;5;202m")
GRY=("\033[01;38;5;242m")
NVD=("\033[1;42;97m")
AMD=("\033[1;41;97m")
RESET=("\033[0m")
LOAD=("\033[1;49;32m")
version=(0.8)
if (os.path.isdir('/data/data/com.termux')):
    OS = ('Termux')
elif ('linux') in (sys.platform):
    OS = ('Linux')
elif ('win') in (sys.platform):
    OS = ('Windows')
script = (os.path.basename(sys.argv[0]))
if ('.py') in script:
    script = (script.split('.')[0])
counter = [0]
currency_options = ['usd', 'gbp', 'cad', 'eur', 'eth']
CHARS = ('/', '-', '\\', '|')
Agent = ('Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36')
def cls():
    os.system('clear')
def loading(duration):	
    for i in range(10):
        counter[0] += 1
        sys.stdout.write('\r')
        sys.stdout.write(f"{LOAD}%s{RESET}" % CHARS[counter[0] % len(CHARS)])
        sys.stdout.flush()
        sleep(duration)
    sys.stdout.write("\b ")
    return()
def echo(data):
    blank = ' '
    s = ''
    for l in blank:
        sys.stdout.write('\r')
        if 'Error' in data:
            sys.stdout.write(f"{DRK}[{RD}■{DRK}] {GRN}{data}")
        else:
            sys.stdout.write(f"{DRK}[{GRN}■{DRK}] {GRN}{data}")
        s += (f"{l}")
        sys.stdout.flush()
        sleep(0.3)
        sys.stdout.write("\n")
def icon():
    icon = (f"""
                   ..
                  {WHT},d{GRY}:.
                 {WHT}:dd{GRY}::,
                {WHT}lddd{GRY}:::;
              {WHT}.odddd{GRY}:::::.
             {WHT};dddddd{GRY}::::::'
            {WHT}cddddddo{GRY}:::::::,
           {WHT}oddool{DRK}c::{GRY},,,;;::::
         {WHT}'olcc{DRK}::::::{GRY},,,,,,,;;:.
            {DRK}::::::::{GRY},,,,,,,,
          {WHT}.    '::::{GRY},,,,.    .
           {WHT}'c'.    :{GRY},     .,.
             {WHT}ddo;.  {GRY}  .';::
              {WHT}'ddddc{GRY},::::.
                {WHT}dddd{GRY}::::
                 {WHT}'dd{GRY}::.
                   {WHT}d{GRY}:    """)
    sys.stdout.write(f"{icon}\n")
def usage():
    info = (f"""{PRP}({RD}Doesn't work with {GRN}GTX {WHT}1050{PNK}/{WHT}1050{GRN}ti {RD}\u2639 {PRP}){DRK}
{GRY}Mining {GRN}Ethereum {GRY}requires {RD}>4GB {GRN}VRAM{PNK}. {GRY}Therefore{PNK}, {GRY}you {RD}cannot {GRY}mine {GRN}ETH {GRY}with a {GRN}1050{PNK}. 
{DRK}Information received from {GRN}{script} {DRK}is thanks to {ORN}https{PNK}://{ORN}hashrate.no
{GRY}Formatting the {GRN}GPU {GRY}in your command{PNK}:
      {NVD}NVIDIA{RESET}{DRK} 
{GRN}3060 {WHT}({GRN}NVIDIA RTX 3060 {PNK}({WHT}Non-LHR{PNK}){WHT})
{GRN}3060lhr {WHT}({GRN}NVIDIA RTX 3060 {PNK}({WHT}LHR{PNK}){WHT})
{GRN}3060ti {WHT}({GRN}NVIDIA RTX 3060 Ti {PNK}({WHT}Non-LHR{PNK}){WHT})
{GRN}3060tilhr {WHT}({GRN}NVIDIA RTX 3060 Ti {PNK}({WHT}LHR{PNK}){WHT})
{GRN}2060s {WHT}({GRN}NVIDIA RTX 2060 Super{WHT})
{PNK}& {GRN}Many more{PNK}. . .
        {AMD}AMD{RESET}{DRK}
{RD}580 {WHT}({RD}RX 580{WHT})
{RD}5600xt {WHT}({RD}RX 5600 XT{WHT})
{RD}vega56 {WHT}({RD}AMD VEGA 56{WHT})
{PNK}& {RD}Many more{PNK}. . .
{GRY}Search mining details on a {GRN}GPU
{PNK}./{GRN}{script} {DRK}[{RD}-g{PNK}/{RD}--gpu {WHT}({GRN}3060 {PNK}| {GRN}3060lhr {PNK}| {GRN}3060ti {PNK}| {GRN}3060tilhr {PNK}| {RD}5600xt {PNK}| {GRN}Etc{WHT}){DRK}]
{PNK}./{GRN}{script} {RD}-c{GRN} 3060
{GRY}List all cards capable of mining {GRN}ETH
{PNK}./{GRN}{script} {RD}-l{PNK}/{RD}--list
{GRY}Now has the ability to convert {GRN}GBP{PNK}, {GRN}CAD {GRY}or {GRN}USD {GRY}to {GRN}ETH {GRY}and vice versa{PNK}.
{GRY}Options{PNK}: 
{GRN}{script} {RD}-c{PNK}/{RD}--currency {WHT}({GRN}eth{PNK}|{GRN}usd{PNK}|{GRN}cad{PNK}|{GRN}gbp{PNK}|{GRN}cad{WHT}) {RD}-i{PNK}/{RD}--into {WHT}({GRN}eth{PNK}|{GRN}usd{PNK}|{GRN}gbp{PNK}|{GRN}cad{PNK}|{GRN}eur{WHT}) {RD}-n{PNK}/{RD}--amount {GRN}10
{GRY}Example syntax to convert {RD}0{PNK}.{RD}2 {GRN}ETH {GRY}into {GRN}GBP{PNK}/{GRN}USD{PNK}/{GRN}CAD{PNK}/{GRN}EUR{PNK}:
{GRN}{script} {RD}-c {GRN}eth {RD}-t {GRN}gbp {RD}-n {GRN}0.2 
{GRY}Or convert {GRN}$175 USD {GRY}into {GRN}ETH{PNK}:
{GRN}{script} {RD}-c {GRN}usd {RD}-t {GRN}eth {RD}-n {GRN}175
 """)
    info = info.replace('\r','.')
    sys.stdout.write(f"{info}\b")
class get():
    def __init__(self) -> None:
        self.session = requests.Session()
        self.session.headers.update({'User-Agent':f"{Agent}"})
        parser = argparse.ArgumentParser(description=f'{GRN}{script} ETH mining stats',add_help=False)
        parser.add_argument('-g', '--gpu', required=False, action=('store'),help=(f'{GRN}GPU to search for'))
        parser.add_argument('-h', '--help',action='help', default=argparse.SUPPRESS,help=(f'Show this help menu'))
        parser.add_argument('-u', '--usage' ,required=False, action=('store_true'), help=(f"Advanced Usage"))
        parser.add_argument('-v', '--version' ,required=False, action=('store_true'), help=(f"{script} Version"))
        parser.add_argument('-l', '--list' ,required=False, action=('store_true'), help=(f"List all cards capable of mining ETH"))
        parser.add_argument('-p', '--price' ,required=False, action=('store_true'), help=(f"Show BTC and ETH price"))
        parser.add_argument('-c', '--convert' ,required=False, action=('store'),type=(str.lower), help=(f"Currency to conver ETH/USD/GBP/CAD/EUR"))
        parser.add_argument('-i', '--into' ,required=False, action=('store'),type=(str.lower), help=(f"Currency to convert into ETH/USD/GBP/CAD/EUR"))
        parser.add_argument('-n', '--amount' ,required=False, action=('store'),type=(float), help=(f"Amount to convert"))
        args = (parser.parse_args())
        if args.price:
            loading(0.05)
            self.get_current_price()
        if args.version:
            icon()
            data = (f"{DRK}Version: {GRN}{script} {PNK}{version} {DRK}[{GRN}■{DRK}]")
            echo(data)
            exit(0);
        try:
            if args.convert:
                to_convert = (args.convert)
                if (args.amount) == (0):
                    data = (f"{RD}Error{PNK}: {DRK}Cannot convert {RD}{args.amount} {DRK}[{RD}■{DRK}]")
                    echo(data)
                    exit(0);
                elif (args.amount) > (0):
                    amount = (args.amount)
                else:
                    data = (f"{RD}Error{PNK}: {DRK}You must supply an amount to convert {DRK}[{RD}■{DRK}]")
                    echo(data)
                    exit(0);
                if to_convert not in ['usd'.casefold(), 'gbp'.casefold(), 'cad'.casefold(), 'eur'.casefold(), 'eth'.casefold()]:
                    data = (f"{RD}Error{PNK}: {RD}{to_convert} {DRK}is not a supported currency{PNK}. {DRK}Acceptable currencies are{PNK}: {GRN}USD{PNK},{GRN}GBP{PNK},{GRN}EUR{PNK},{GRN}CAD")
                    echo(data)
                    exit(0);
                if len(args.convert) != (3):
                    data = (f"{RD}Error {PNK}'{RD}{args.convert}{PNK}' {DRK}not recognized{PNK}: {RD}Options are{PNK}: {GRN}eth{PNK}, {GRN}usd{PNK}, {GRN}gbp{PNK},{GRN}cad{PNK},{GRN} eur {DRK}[{RD}■{DRK}]")
                    echo(data)
                if ('eth') in (to_convert):
                    currency = ('ethereum')
                elif 'usd' in (to_convert):
                    currency = ('usd')
                elif ('gbp') in (to_convert):
                    currency = ('gbp')
                elif ('eur') in (to_convert):
                    currency = ('eur')
                elif ('cad') in (to_convert):
                    currency = ('cad')
                if len(args.into) <= (2):
                    data = (f"{RD}Error {PNK}'{RD}{args.into}{PNK}' {DRK}not recognized{PNK}: {RD}Options are{PNK}: {GRN}eth{PNK}, {GRN}usd{PNK}, {GRN}gbp{PNK},{GRN}cad{PNK}, {GRN}eur {DRK}[{RD}■{DRK}]")
                    echo(data)
                else:
                    convert_to = (args.into)
                    if convert_to not in ['usd'.casefold(), 'gbp'.casefold(), 'cad'.casefold(), 'eur'.casefold(), 'eth'.casefold()]:
                        data = (f"{RD}Error{PNK}: {RD}{to_convert} {DRK}is not a supported currency{PNK}. {DRK}Acceptable currencies are{PNK}: {GRN}USD{PNK},{GRN}GBP{PNK},{GRN}EUR{PNK},{GRN}CAD")
                        echo(data)
                        exit(0);
                loading(0.05)
                self.convert(currency,amount,convert_to)
        except:
            pass
        if args.list:
            data = (f"{DRK}Stats brought to you by{PNK}: {GRN}https{DRK}://{GRN}hashrate.no {DRK}[{GRN}■{DRK}]")
            echo(data)
            sleep(1);
            icon()
            self.list_cards()
            exit(0);
        if args.usage:
            usage()
            exit(0);
        else:
            pass
        if args.gpu:
            if args.gpu == ('1050') or args.gpu == ('1050ti'):
                data = (f"{RD}Error{PNK}: {RD}\u2639 {DRK} The {GRN}GTX {GRN}1050{PNK}/{WHT}1050{GRN}ti {DRK}cannot be used to mine {GRN}ETH {RD}\u2639  {DRK}[{RD}■{DRK}]")
                echo(data)
                data = (f"{RD}Error{PNK}: {GRY}Mining {GRN}ETH {GRY}requires {RD}>4GB {GRN}VRAM {DRK}[{RD}■{DRK}]")
                echo(data)
                exit(0)
            else:
                data = (f"{DRK}Stats brought to you by{PNK}: {GRN}https{DRK}://{GRN}hashrate.no {DRK}[{GRN}■{DRK}]")
                echo(data)
                sleep(1);
                self.card = (args.gpu)
                icon()
                self.get_mhs()
        else:
            pass
    def get_gas_price(self):
        data = (f"Maybe a 'gas price' here ?")
        echo(data)
    def get_current_price(self):
        btc_page = ('https://api.coindesk.com/v1/bpi/currentprice/usd.json')
        page = (f'https://www.hashrate.no/')
        page2 = (f"https://walletinvestor.com/converter/ethereum/usd/1")
        retrieve = (self.session.get(page).text)
        get_btc = (self.session.get(btc_page).text)
        retrieve2 = (self.session.get(page2).text)
        try:
            btc_price = (re.findall('"rate":"(.*?)"',get_btc)[0])
        except:
            btc_price = (f'N/A')
        try:
            eth_price = (re.findall("ETH: <b>(.*?)</", retrieve)[0])
        except:
            eth_price = ('N/A')
        try:
            down = (re.findall('glyphicon-menu-down"></i> (.*?)</',retrieve2)[0])
            down_clean = (down.replace(" ", ""))
        except:
            down = (re.findall('glyphicon-menu-up"></i> (.*?)</',retrieve2)[0])
            down_clean = (down.replace(" ", ""))
        if ('-') in down:
            data = (f"{DRK}BTC Price{PNK}: {GRN}${GRN}{btc_price}")
            echo(data)
            loading(0.01)
            data = (f"{DRK}ETH Price{PNK}: {GRN}{eth_price} {PNK}.::. {DRK}Down{PNK}: {RD}{down_clean}\r")
            echo(data)
        else:
            data = (f"{DRK}BTC Price{PNK}: {GRN}${GRN}{btc_price}")
            echo(data)
            loading(0.01)
            data = (f"{DRK}ETH Price{PNK}: {GRN}{eth_price} {PNK}.::. {DRK}Up{PNK}: {GRN}{down_clean}\r")
            echo(data)
    def convert(self,currency,amount,convert_to):
        if convert_to == 'gbp':
            convert_symbol = ('£')
        elif convert_to == 'usd':
            convert_symbol = ('$')
        elif convert_to == 'cad':
            convert_symbol = ('$')
        elif convert_to == 'eur':
            convert_symbol = ('€')
        elif convert_to == 'eth':
            convert_symbol = ('♦')
        if convert_to == ('eth'):
            convert_to = ('ethereum')
        if currency == ('ethereum'):
            page = (f"https://walletinvestor.com/converter/{currency}/{convert_to}/{amount}")
            retrieve = (self.session.get(page).text)
            raw = (re.findall('converter-title-amount">(.*?)</span>',retrieve)[0])
            converted = (f"{int(float(raw)):.2f}")
            if currency == 'gbp':
                currency_symbol = ('£')
            elif currency == 'usd':
                currency_symbol = ('$')
            elif currency == 'cad':
                currency_symbol = ('$')
            elif currency == 'eur':
                currency_symbol = ('€')
            elif currency == 'ethereum':
                currency_symbol = ('♦')
            loading(0.05)
            data = (f"{GRY}Converting{PNK}: {GRN}{currency_symbol}{amount} {currency.upper()[0:3]} {GRY}into{PNK}: {GRN}{convert_symbol}{converted} {convert_to.upper()} {DRK}[{GRN}■{DRK}]")
            echo(data)
        elif currency == ('usd') or currency == ('gbp') or currency == ('cad') or currency == ('eur'):
            page = (f"https://walletinvestor.com/converter/{currency}/{convert_to}/{amount}")
            retrieve = (self.session.get(page).text)
            raw = (re.findall('converter-title-amount">(.*?)</span>',retrieve)[0])
            converted = (f"{int(float(raw)):.2f}")
            if currency == 'gbp':
                currency_symbol = ('£')
            elif currency == 'usd':
                currency_symbol = ('$')
            elif currency == 'cad':
                currency_symbol = ('$')
            elif currency == 'eur':
                currency_symbol = ('€')
            elif currency == 'ethereum':
                currency_symbol = ('♦')
            loading(0.05)
            data = (f"{GRY}Converting{PNK}: {GRN}{currency_symbol}{amount} {currency.upper()} {GRY}into{PNK}: {GRN}{convert_symbol}{raw} {convert_to.upper()[0:3]} {DRK}[{GRN}■{DRK}]")
            echo(data)
    def get_mhs(self):
        page2 = (f"https://walletinvestor.com/converter/ethereum/usd/1")
        btc_page = ('https://api.coindesk.com/v1/bpi/currentprice/usd.json')
        get_btc = (self.session.get(btc_page).text)
        page = (f'https://www.hashrate.no/{self.card}')
        retrieve = (self.session.get(page).text)
        retrieve2 = (self.session.get(page2).text)
        try:
            btc_price = (re.findall('"rate":"(.*?)"',get_btc)[0])
        except:
            btc_price = (f'N/A')
        try:
            eth_price = (re.findall("ETH: <b>(.*?)</", retrieve)[0])
        except:
            eth_price = ('N/A')
        try:
            gpu = (re.findall("<title>(.*?) -",retrieve)[0])
        except:
            gpu = (f'N/A')
        try:
            eth_rate = (re.findall("ETH</a></font></td><td>(.*?)</", retrieve)[0])
        except:
            eth_rate = ('N/A')
        try:
            power = (re.findall(f"{eth_rate}</td><td><b>(.*?)</", retrieve)[0])
            power2 = (re.findall(f"{power}</b><br />(.*?)</", retrieve)[0])
        except:
            power = ('N/A')
            power2 = ('N/A')
        try:
            efficiency = (re.findall(f"{power}</b><br />{power2}</td><td>(.*?)</", retrieve)[0])
        except:
            efficiency = ('N/A')
        try:
            oneday = (re.findall(f"{power2}</td><td>{efficiency}</td><td><b>(.*?)</", retrieve)[0])
            oned = oneday.rsplit('$')[1]
            weekly = (D(oned) * 7)
            monthly = (D(oned) * 30)
        except:
            oneday = ('N/A')
            weekly = ('N/A')
            monthly = ('N/A')
        try:
            oneday2 = (re.findall(f"{efficiency}</td><td><b>{oneday}</b><br />(.*?)</", retrieve)[0])
        except:
            oneday2 = ('N/A')
        try:
            rvn_rate = (re.findall("RVN</a></font></td><td>(.*?)</", retrieve)[0])
        except:
            rvn_rate = ('N/A')
        try:
            msrp = (re.findall("MSRP: (.*?)Released:", retrieve)[0])
        except:
            msrp = ('N/A')
        try:
            release_date = (re.findall("Released:(.*?)TDP:", retrieve)[0])
        except:
            release_date = ('N/A')
        try:
            tdp = (re.findall("TDP: </td><td>(.*?)</", retrieve)[0])
        except:
            tdp = ('N/A')
        try:
            test = (oneday.split('.')[1])
            roi = (re.findall(f"{test}</b></td><td>(.*?)</td", retrieve)[0])
        except:
            roi = ('N/A')
        try:
            if "NVIDIA" in gpu:
                gpu_name = (f"{NVD}{gpu}{RESET} {DRK}[{GRN}■{DRK}]")
                data = (f"{gpu_name}")
                echo(data)
                loading(.2)
                try:
                    down = (re.findall('glyphicon-menu-down"></i> (.*?)</',retrieve2)[0])
                    down_clean = (down.replace(" ", ""))
                except:
                    down = (re.findall('glyphicon-menu-up"></i> (.*?)</',retrieve2)[0])
                    down_clean = (down.replace(" ", ""))
                if ('-') in down:
                    data = (f"{DRK}BTC Price{PNK}: {GRN}${GRN}{btc_price} {PNK}.::. {DRK}ETH Price{PNK}: {GRN}{eth_price} {PNK}.::. {DRK}Down{PNK}: {RD}{down_clean}\r")
                    echo(data)
                else:
                    data = (f"{DRK}BTC Price{PNK}: {GRN}${GRN}{btc_price} {PNK}.::. {DRK}ETH Price{PNK}: {GRN}{eth_price} {PNK}.::. {DRK}Up{PNK}: {GRN}{down_clean}")
                    echo(data)
                loading(0.05)
                data = (f"{GRN}ETH {DRK}Hashrate{PNK}({GRN}DaggerHashimoto{PNK}){PNK}: {GRN}{eth_rate} {PNK}") 
                echo(data)
                loading(0.05)
                data = (f"{DRK}24h{PNK}: {GRN}{oneday}{PNK} / {DRK}Weekly{PNK}: {GRN}${weekly}{PNK} / {DRK}Monthly{PNK}: {GRN}${monthly}")
                echo(data)
                loading(0.05)
                data = (f"{DRK}Released{PNK}: {GRN}{release_date} {PNK}.::. {DRK}MSRP{PNK}: {GRN}{msrp} ")
                echo(data)
                loading(0.05)
                data = (f"{DRK}ROI{PNK}: {GRN}{roi} {PNK}.::. {DRK}Efficiency{PNK} {GRN}{efficiency}")
                echo(data)
                loading(0.05)
                data = (f"{DRK}TDP{PNK}: {GRN}{tdp} {PNK}.::. {DRK}Power{PNK}: {GRN}{power} {PNK}:: {GRN}{power2}")
                echo(data)          
            elif "AMD" in gpu:
                gpu_name = (f"{AMD}{gpu}{RESET} {DRK}[{GRN}■{DRK}]")
                data = (f"{gpu_name}")
                echo(data)
                loading(.2)
                try:
                    down = (re.findall('glyphicon-menu-down"></i> (.*?)</',retrieve2)[0])
                    down_clean = (down.replace(" ", ""))
                except:
                    down = (re.findall('glyphicon-menu-up"></i> (.*?)</',retrieve2)[0])
                    down_clean = (down.replace(" ", ""))
                if ('-') in down:
                    data = (f"{DRK}BTC Price{PNK}: {GRN}${RD}{btc_price} {PNK}.::. {DRK}ETH Price{PNK}: {RD}{eth_price} {PNK}.::. {DRK}Down{PNK}: {RD}{down_clean}\r")
                    echo(data)
                else:
                    data = (f"{DRK}BTC Price{PNK}: {RD}${btc_price} {PNK}.::. {DRK}ETH Price{PNK}: {RD}{eth_price} {PNK}.::. {DRK}Up{PNK}: {GRN}{down_clean}")
                    echo(data)
                loading(0.05)
                data = (f"{RD}ETH {DRK}Hashrate{PNK}({RD}DaggerHashimoto{PNK}){PNK}: {RD}{eth_rate} {PNK}") 
                echo(data)
                loading(0.05)
                data = (f"{DRK}24h{PNK}: {RD}{oneday}{PNK}/{DRK}Weekly{PNK}: {RD}${weekly}{PNK}/{DRK}Monthly{PNK}: {RD}${monthly}")
                echo(data)
                loading(0.05)
                data = (f"{DRK}Released{PNK}: {RD}{release_date} {PNK}.::. {DRK}MSRP{PNK}: {RD}{msrp} ")
                echo(data)
                loading(0.05)
                data = (f"{DRK}ROI{PNK}: {RD}{roi} {PNK}.::. {DRK}Efficiency{PNK} {RD}{efficiency}")
                echo(data)
                loading(0.05)
                data = (f"{DRK}TDP{PNK}: {RD}{tdp} {PNK}.::. {DRK}Power{PNK}: {RD}{power}")
                echo(data)
        except KeyboardInterrupt:
            data = (f"{RD}Error{PNK}: {DRK}Keyboard Interruption {DRK}[{RD}■{DRK}]")
            echo(data)
            exit(0);
            
    def list_cards(self):
        page = (f'https://www.hashrate.no/')
        retrieve = (self.session.get(page).text)
        all_cards = (list(set(re.findall(f"href=/(.*?)>", retrieve))))
        loading(.2)
        try:
            for a in all_cards:
                all_cards_clean = (list(set(re.findall(f"href=/{a}>(.*?)</a>", retrieve))))
                for card in all_cards_clean:
                    if "NVIDIA" in card:
                        data = (f"{DRK}Card name{PNK}: {NVD}{card}{RESET} {PNK}.:|:. {DRK}Command name{PNK}: {GRN}{a}")
                        echo(data)
                        loading(0.02)
                    elif "AMD" in card:
                        data = (f"{DRK}Card name{PNK}: {AMD}{card}{RESET} {PNK}.:|:. {DRK}Command name{PNK}: {RD}{a}")
                        echo(data)
                        loading(0.02)
                    else:
                        data = (f"{DRK}Card name{PNK}: {GRN}{card} {PNK}.:|:. {DRK}Command name{PNK}: {GRN}{a}")
                        echo(data)
                        loading(0.02)
        except KeyboardInterrupt:
            data = (f"{RD}Error{PNK}: {DRK}Keyboard Interruption {DRK}[{RD}■{DRK}]")
            echo(data)
            exit(0);
get()
